### ISSUE TITLE
Separate "Algorithm" and "Mode" into distinct properties?

### ISSUE THREAD

**Issue by celic** [usnistgov/ACVP#167]
We have

```
{
    "algorithm": "AES-GCM",
    capabilities
}
```

or 

``` 
{
    "algorithm": "SHA-256",
    capabilities
}
```

which doesn't match up well to the validation side of the system. For RSA, we have

```
{
    "algorithm": "RSA",
    "algSpecs": [{
        "modeSpecs": {
            "mode": "keyGen",
            "capSpecs": [{
                "capabilities": "all the fields"
             }]
         }
    },
    {
        "modeSpecs": {
            "mode": "sigVer",
            "capSpecs": [{
                "capabilities": "all the fields"
             }]
         }
    }]
}
```

with the mode defined within the algorithm. I'm not exactly sure how it would look in terms of the symmetric ciphers and hash functions, but should the mode be a separate entity from the algorithm for consistency?

**Comment by atvassilev** [comment]
Concerning block ciphers, useful examples are provided by the SP 800-38[A-E] series of specifications. Each document talks about an underlying block cipher algorithm and a mode that goes with it. By the notion established in these standards AES-ECB is a mode of the AES algorithm, not an algorithm. Similarly for the other block cipher modes and algorithms. 

In my opinion, this is a logical way of thinking about algorithms and modes that we should/could adopt in our specifications. 

The same approach could be used for hash functions: three underlying algorithms: SHA-1, SHA-2 and SHA-3 with modes 224, 256, 384, 512, 512/256.

**Comment by sam1689** [comment]
Hi Guys,
    I am not sure what is being proposed exactly - but I am
      definitely against changing the way we handle the previously
      implemented algorithms like AES, DES, and SHA...
    I do not think we can get around needing to use a "loose"
      definition for "algorithm" where sometimes it is more exactly
      referring to an algorithm proper, and sometimes it is more loosely
      representing a mode of a broader algorithm... 
    
    If we were to go with something more generic like "name" that
      would lose value and communicate less... 
    
    If we were to change that to "AlgorithmMode" or "algMode" it
      would still have to be stretched in such cases as it actually
      refers to something more than a mode...
    I vote that we put up with the "stretching" of the field name
      "algorithm" so that we can provide some consistency in how we
      parse and process communication to/from the clients.
    I was thinking of something like this for RSA Sig Gen, Key Gen,
      and Sig Verify...
    {
       "capabilityExchange": {
             "algorithms": [
             {
              "algorithm": "RSA-SigGen",
                  "prereqVals": [
                      {"algorithm": "DRBG", "value": "same"}, 
                      {"algorithm": "SHA", "value": "same"}
                      ],
                  "algSpecs" : [
                      {
                       "sigType": "X9.31",
                       "options" : [
                           {"mod" : 2048, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"]},
                           {"mod" : 3072, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"]},
                           {"mod" : 4096, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"]}
                       ]
                      },
                      {
                       "sigType" : "PKCS1v1.5",
                       "options" : [
                           {"mod" : 2048, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"]},
                           {"mod" : 3072, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"]},
                           {"mod" : 4096, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"]}
                       ]
                      },
                      {
                       "sigType" : "PKCS1PSS",
                       "options" : [
                           {"mod" : 2048, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"], "salt" : [28, 32, 64]},
                           {"mod" : 3072, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"], "salt" : [28, 32, 64]},
                           {"mod" : 4096, "hash" : ["SHA-224",
      "SHA-256", "SHA-512"], "salt" : [28, 32, 64]}
                       ]
                     }
                 ]
              }
           ]
         }
      }
    It has some simplification of the json but I think it captures
      all the necessary parameters....
    Thoughts?
    
    Cheers,
    Sam
    
    
    On 6/7/17 4:46 PM, Apostol Vassilev
      wrote:
    
    
      Concerning block ciphers, useful examples are provided by the
        SP 800-38[A-E] series of specifications. Each document talks
        about an underlying block cipher algorithm and a mode that goes
        with it. By the notion established in these standards AES-ECB is
        a mode of the AES algorithm, not an algorithm. Similarly for the
        other block cipher modes and algorithms.
      In my opinion, this is a logical way of thinking about
        algorithms and modes that we should/could adopt in our
        specifications.
      The same approach could be used for hash functions: three
        underlying algorithms: SHA-1, SHA-2 and SHA-3 with modes 224,
        256, 384, 512, 512/256.
      —
        You are receiving this because you are subscribed to this
        thread.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/usnistgov/ACVP","title":"usnistgov/ACVP","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/usnistgov/ACVP"}},"updates":{"snippets":[{"icon":"PERSON","message":"@atvassilev in #167: Concerning block ciphers, useful examples are provided by the SP 800-38[A-E] series of specifications. Each document talks about an underlying block cipher algorithm and a mode that goes with it. By the notion established in these standards AES-ECB is a mode of the AES algorithm, not an algorithm. Similarly for the other block cipher modes and algorithms. \r\n\r\nIn my opinion, this is a logical way of thinking about algorithms and modes that we should/could adopt in our specifications. \r\n\r\nThe same approach could be used for hash functions: three underlying algorithms: SHA-1, SHA-2 and SHA-3 with modes 224, 256, 384, 512, 512/256.\r\n "}],"action":{"name":"View Issue","url":"https://github.com/usnistgov/ACVP/issues/167#issuecomment-306919687"}}}

**Comment by atvassilev** [comment]
Have we settled this issue?

**Comment by sam1689** [comment]
Hi Apostol,

I still don't see much value in the extra "modeSpecs" layer.  But of course it is code - we can make it whatever we want to that fits the json guides :-)

This is an example of what we have on the client side right now:

{
  "algorithm":"RSA",
  "prereqVals":[],
  "algSpecs":[
    {
     "sigGen":[
       {
        "sigType":"X9.31",
        "capSigType":[]
       },
       {
        "sigType":"PKCS1PSS",
        "capSigType":[]
       }
      ]
    }
  ]
}
The long version looks like this:
{
  "algorithm":"RSA",
  "prereqVals":[{"algorithm":"DRBG", "valValue":"same"}, {"algorithm":"SHA", "valValue":"same"}],
  "algSpecs":[
    {
     "sigGen":[
       {
        "sigType":"X9.31",
        "capSigType":[
          {
            "modulo":2048,
            "hashAlg":["SHA-224", "SHA-256", "SHA-512"]
           },
           {
            "modulo":3072,
            "hashAlg":["SHA-224", "SHA-256", "SHA-512"]
          },
          {
            "modulo":4096,
            "hashAlg":["SHA-224","SHA-256","SH A-512"]
            }
       ]
     },
    {
       "sigType":"PKCS1v1.5",
       "capSigType":[
         {
          "modulo":2048,
          "hashAlg":["SHA-224","SHA-256","SHA-512"]
         },
         {
          "modulo":3072,
          "hashAlg":[ "SHA-224", "SHA-256", "SHA-512"]
         },
         {
          "modulo":4096,
          "hashAlg":[ "SHA-224", "SHA-256", "SH A-512"]
         }
        ]
      },
     {
      "sigType":"PKCS1PSS",
      "capSigType":[
        {
         "modulo":2048,
         "hashAlg":["SHA-224","SHA-256","SHA-512"],
         "saltLen":[28, 32, 64]
        },
        {
         "modulo":3072,
         "hashAlg":["SHA-224","SHA-256","SHA-512"],
         "saltLen":[28,32,64]
        },
        {
         "modulo":4096,
         "hashAlg":["SHA-224","SHA-256","SHA-512"],
         "saltLen":[28,32,64]
        }
      ]
     }
   ]
  }
 ]
}

**Comment by sam1689** [comment]
For some reason the comment editor removes the spacing and indentation :-(

**Comment by Kritner** [comment]
@sam1689 are you using the markdown code wrapping?

```
{
	"algorithm": "RSA",
	"prereqVals": [{
		"algorithm": "DRBG",
		"valValue": "same"
	}, {
		"algorithm": "SHA",
		"valValue": "same"
	}],
	"algSpecs": [{
		"sigGen": [{
				"sigType": "X9.31",
				"capSigType": [{
						"modulo": 2048,
						"hashAlg": ["SHA-224", "SHA-256", "SHA-512"]
					},
					{
						"modulo": 3072,
						"hashAlg": ["SHA-224", "SHA-256", "SHA-512"]
					},
					{
						"modulo": 4096,
						"hashAlg": ["SHA-224", "SHA-256", "SH A-512"]
					}
				]
			},
			{
				"sigType": "PKCS1v1.5",
				"capSigType": [{
						"modulo": 2048,
						"hashAlg": ["SHA-224", "SHA-256", "SHA-512"]
					},
					{
						"modulo": 3072,
						"hashAlg": ["SHA-224", "SHA-256", "SHA-512"]
					},
					{
						"modulo": 4096,
						"hashAlg": ["SHA-224", "SHA-256", "SH A-512"]
					}
				]
			},
			{
				"sigType": "PKCS1PSS",
				"capSigType": [{
						"modulo": 2048,
						"hashAlg": ["SHA-224", "SHA-256", "SHA-512"],
						"saltLen": [28, 32, 64]
					},
					{
						"modulo": 3072,
						"hashAlg": ["SHA-224", "SHA-256", "SHA-512"],
						"saltLen": [28, 32, 64]
					},
					{
						"modulo": 4096,
						"hashAlg": ["SHA-224", "SHA-256", "SHA-512"],
						"saltLen": [28, 32, 64]
					}
				]
			}
		]
	}]
}
```

it's three backticks at the beginning and end of a block (or highlight the text and use the "<>" button at the top of the editor.

**Comment by sam1689** [comment]
Didn't know that - thanks Russ!

**Comment by celic** [comment]
This was addressed with #198 and the issue will be closed. As RSA reaches integration if there are any questions the issue can be reopened.
